fix: include type definitions when publishing

achou11 opened the pull request:
Mainly:

- Sets up type defs generation
- Updates npm package setup to include generated types
- ~**BREAKING CHANGE**: renames the exports to include the `.js` suffix at the end, which I believe is generally recommended/idiomatic. Not strictly necessary for us to do so open to reverting that change if preferred~ EDIT: reverted

Should introduce `noUncheckedIndexedAccess` in the tsconfig as a follow-up but it will lead to errors that will require code changes outside the scope of this PR.

achou11 commented on lib/utils/streams.js:
Changes here related to `TransformOptions` are not ideal but they were causing an error when emitting declarations via `npm run build:types`. 

Error was:

```ts
lib/utils/streams.js:1:1 - error TS9006: Declaration emit for this file requires using private name '_IReadable' from module '"/Users/andrewchou/GitHub/digidem/styled-map-package/node_modules/@types/readable-stream/index"'. An explicit type annotation may unblock declaration emit.

1 import { Readable, Writable, Transform } from 'readable-stream'
```

achou11 replied:
Fixed via 31213d05aedd4d92caa53a1f0f58487a6e8200f0

achou11 commented on package.json:
```diff
@@ -80,6 +89,7 @@
     "png-stream": "^1.0.5",
     "prettier": "^3.3.3",
     "random-bytes-readable-stream": "^3.0.0",
+    "rimraf": "^4.4.1",
```
Pinned to v4 since >= v5 removes support for node 18

achou11 commented on package.json:
```diff
@@ -16,10 +16,19 @@
     "smp": "./bin/smp.js"
   },
   "type": "module",
+  "files": [
+    "bin",
+    "dist",
+    "lib",
+    "map-viewer"
+  ],
```
not sure about including `map-viewer`. anything else to include here otherwise?

achou11 replied:
`map-viewer` makes sense to include since one of the bin scripts uses it. still an open question about anything else that should be added to this list...

gmaclennan replied:
think that should be it

achou11 commented on package.json:
```diff
@@ -3,23 +3,54 @@
-    "./style-downloader": "./lib/style-downloader.js",
-    "./tile-downloader": "./lib/tile-downloader.js",
-    "./download": "./lib/download.js"
+    ".": {
+      "types": "./dist/index.d.ts",
+      "import": "./lib/index.js"
+    },
+    "./reader.js": {
+      "types": "./dist/reader.d.ts",
+      "import": "./lib/reader.js"
+    },
+    "./writer.js": {
+      "types": "./dist/writer.d.ts",
+      "import": "./lib/writer.js"
+    },
+    "./server.js": {
+      "types": "./dist/server.d.ts",
+      "import": "./lib/server.js"
+    },
+    "./style-downloader.js": {
+      "types": "./dist/style-downloader.d.ts",
+      "import": "./lib/style-downloader.js"
+    },
+    "./tile-downloader.js": {
+      "types": "./dist/tile-downloader.d.ts",
+      "import": "./lib/tile-downloader.js"
+    },
+    "./download.js": {
+      "types": "./dist/download.d.ts",
+      "import": "./lib/download.js"
+    }
```
changes here informed by [publint](https://publint.dev/rules#types_not_exported), which is generally a super helpful tool that i've become reliant on

gmaclennan reviewed (changes requested):
great, thanks for this! Removing `TransformOptions` to satisfy typescript bothered me, so I fixed that with 31213d05aedd4d92caa53a1f0f58487a6e8200f0

I don't agree with the change to add `.js` to exports. Most modules I have used that define exports do not include an extension, including node modules (e.g. `fs/promises`), and I think it leads to some confusion because it suggests a file is being imported. The file extension is required by ESM when importing a local file, but that is a different thing.

achou11 commented:
> Removing `TransformOptions` to satisfy typescript bothered me, so I fixed that with [31213d0](https://github.com/digidem/styled-map-package/commit/31213d05aedd4d92caa53a1f0f58487a6e8200f0)

nice! I tried something almost similar to that before and it didn't work. glad you fixed it.

> I don't agree with the change to add `.js` to exports. Most modules I have used that define exports do not include an extension, including node modules (e.g. `fs/promises`), and I think it leads to some confusion because it suggests a file is being imported. The file extension is required by ESM when importing a local file, but that is a different thing.

Yeah that makes sense. Reverted via 2df05b9312ffc671095bfb472643bf9f0e6aa04c

achou11 commented:
One last thing to potentially resolve: what's the best way to expose the types exported from `lib/types.ts`? e.g. needing access to `SMPStyle`

gmaclennan commented:
> what's the best way to expose the types exported from `lib/types.ts`?

you can add this line to `index.js`:

```
/** @typedef {import('./types.js').SMPStyle} SMPStyle */
```

Then you can do:

```ts
import type { SMPStyle } from 'styled-map-package'
```

You could do the same in any of the other export entry points.

gmaclennan reviewed (approved):
looks good to me!

achou11 commented:
> > what's the best way to expose the types exported from `lib/types.ts`?
> 
> you can add this line to `index.js`:
> 
> ```
> /** @typedef {import('./types.js').SMPStyle} SMPStyle */
> ```
> 
> Then you can do:
> 
> ```ts
> import type { SMPStyle } from 'styled-map-package'
> ```
> 
> You could do the same in any of the other export entry points.

Awesome, seems to work! (see 05084f9b509bf5ec423b398b70a1e89f9f5f1c0c)
